050-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,680 @@
+>     status:
+>       observedGeneration: 124 # == generation => it is not syncrhonized
+>     ```
+> 
+> 3. The operator notices the `metadata.deletionTimestamp` field.
+>    1. If there is more than one `KafkaTopic` which is trying to manage this topic then deletion from Kafka is not attempted;
+>    the operator will remove the finalizer and Kube will remove the resource; upon timed reconciliation the condition on the other `KafkaTopic` will be removed and reconciliation via that other `KafkaTopic` will proceed as normal.
+>    2. If deletion is disallowed via `delete.topic.enable=false` broker configuration then the operator will remove the finalizer and Kube will remove the resource. The topic in Kafka will thus become unmanaged.
+>    3. Otherwise, the operator attempted to deletes the topic from Kafka.
+>        1. If deletion succeeds the operator will remove `strimzi.io/topic-operator` from the `metadata.finalizer`.
+>        2. If deletion failed with `UNKNOWN_TOPIC_OR_PARTITION` (i.e. the topic didn't exist) the operator will continue to remove the finalize. I.e. this case is not treated as an error.
+>        3. If deletion failed with any other error this is reported via a condition in the `status.conditions` and the finalizer is not removed.
+---
+
+Similarly to how it is possible to "manage" an existing unmanaged topic in Kafka, we will also provide a mechanism for "unmanaging" a managed topic, via a `spec.managed` flag.
+This will allow a `KafkaTopic` to be deleted **without** the topic in Kafka being deleted.
+This can be useful operationally, for example to change the `metadata.name` of a `KafkaTopic`.
+
+> An alternative to `spec.managed` would be "explicit deletion" via a `spec.delete` flag.
+> This alternative was rejected because it's incompatible with gitops.
+
+---
+
+#### Example 4: `KafkaTopic` deletion WITHOUT propagation to Kafka
+
+> 1. Assume the topic is currently Kube-managed
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 123
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       # ...
+>     status:
+>       observedGeneration: 123 # == generation => it is Kube-managed
+>     ```
+> 
+> 2. The user changes the new `spec.managed` field to `False` so that the topic is no longer Kube-managed.
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 124
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       managed: False # do not synchronize to Kafka
+>     status:
+>       observedGeneration: 123 # != generation => it is Kube-managed
+>     ```
+> 
+> 3. The user waits until the operator has observed that change, indicated by the `status.observedGeneration` matching the `metadata.generation`:
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 124
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       managed: False # do not synchronize to Kafka
+>     status:
```

Comment:
We should also remove the `spec.topicName` to help indicate that there might not be a topic in Kafka for this `KafkaTopic`.

Reply:
You cannot do that, because if you remove `.spec.topicName` you might not later know what the actual topic name was, or?

Reply:
That would be fine, actually. 

1. `spec.managed=false` 
2. then changing the `spec.topicName` 
3. then `spec.managed=true`

 just means you changed which topic in Kafka is being managed by the resource, just like it would if step 2 was deletion and recreation.

Reply:
Yes. But the point is that:
* You might fight with some GitOps tooling which would set the `spec.topicName` back
* Assuming the GitOps tooling did not reverse your change, the user will be unable to just edit the spec and set `spec.managed=true` to recreate the topic / have it managed again.

The procedure you describe is fine if done by the user. But I do not think the UTO should edit the `.spec` this way.